### ISSUE TITLE
error: metadata-generation-failed under Python 3.10, Windows 10

### ISSUE THREAD

**Issue by lindemann09** [expyriment/expyriment#198]
_I received this from the technician in the lab in Rotterdam:_

It seems Expyriment does not work under the latest python version (3.10). It seems to needs a 32bit Python v3.7.5. After changing to that, it worked

We get the following error during `pip install expyriment`. Maybe you know a solution?

```
Collecting pygame<2,>=1.9

  Using cached pygame-1.9.6.tar.gz (3.2 MB)

  Preparing metadata (setup.py) ... error

  error: subprocess-exited-with-error

 

  × python setup.py egg_info did not run successfully.

  │ exit code: 1

  ╰─> [18 lines of output]

 

 

      WARNING, No "Setup" File Exists, Running "buildconfig/config.py"

      Using WINDOWS configuration...

 

 

      Download prebuilts to "prebuilt_downloads" and copy to "./prebuilt-x64"? [Y/n]Traceback (most recent call last):

        File "<string>", line 2, in <module>

        File "<pip-setuptools-caller>", line 34, in <module>

        File "D:\Users\ebluser\AppData\Local\Temp\pip-install-f2hgjbtc\pygame_dc2203f23877437a871f62e2e359f04a\setup.py", line 194, in <module>

          buildconfig.config.main(AUTO_CONFIG)

        File "D:\Users\ebluser\AppData\Local\Temp\pip-install-f2hgjbtc\pygame_dc2203f23877437a871f62e2e359f04a\buildconfig\config.py", line 210, in main

          deps = CFG.main(**kwds)

        File "D:\Users\ebluser\AppData\Local\Temp\pip-install-f2hgjbtc\pygame_dc2203f23877437a871f62e2e359f04a\buildconfig\config_win.py", line 576, in main

          and download_win_prebuilt.ask(**download_kwargs):

        File "D:\Users\ebluser\AppData\Local\Temp\pip-install-f2hgjbtc\pygame_dc2203f23877437a871f62e2e359f04a\buildconfig\download_win_prebuilt.py", line 302, in ask

          reply = raw_input(

      EOFError: EOF when reading a line

      [end of output]

 

  note: This error originates from a subprocess, and is likely not a problem with pip.

error: metadata-generation-failed

 

× Encountered error while generating package metadata.

╰─> See above for output.

 

note: This is an issue with the package mentioned above, not pip.

hint: See above for details.
```

**Comment by fladd** [comment]
Expyriment works fine with Python 3.10. This is just another instance of Pygame 1.9 not being made available as wheels on Pypi anymore and hence Pip wants to build it from source. The error seems to indicate that this did not work, probably due to missing tooling. However, I am not sure what exact tooling is needed under Windows to compile Pygame.

Anyway, once we switch to Pygame 2 in Expyriment 0.11.0, Pip can just use the corresponding wheels on PyPi and compiling won't be necessary anymore.

**Comment by lindemann09** [comment]
yes, but need to happen really soon.

**Comment by chrplr** [comment]
Under Ubuntu 20.04, I used to be able to install expyriment with pygame2 by doing a `pip install .` on the source code from github (rather than a `pip install epxyriment` which tries to setup pygame1.9.6).

On Ubuntu 22.04, the newest LTS distrib, `pip install .` (or `python setup.py install`) on the source code crashes with an odd message about install_lib from the setuptools (see below). I am using the same anaoconda pyton 3.8 though and I doubt this is due to my setuptools/distutils install because I have no problem installing other python packages from source. I am banging my head on this issue, which totally prevents me from installing expyriment on Ubuntu 22.04. 

(Also, under ubuntu 22.04 although there is a compatibility package, libsdl1.2-compat_shim that attempts to replace libsld1.2 by hooks to libsdl2, I have issues when trying to compile pygame1.9.6 with it). 

I wish I could help you port expyriment to make it work by default with pygame2 (my understanding is you used both pygame and pyopengl to allow for blocking on vsync, but in pygame2, pygame.display.set_mode() has a `vsync` flag which may allow you to simplify your code (?)). 

Just let me know if I can help in any way.

  
---
    Traceback (most recent call last):
      File "/tmp/pip-req-build-qev55xfx/setup.py", line 357, in <module>
        run()
      File "/tmp/pip-req-build-qev55xfx/setup.py", line 306, in run
        setup(name='expyriment',
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/__init__.py", line 87, in setup
        return distutils.core.setup(**attrs)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 148, in setup
        return run_commands(dist)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 163, in run_commands
        dist.run_commands()
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 967, in run_commands
        self.run_command(cmd)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/dist.py", line 1214, in run_command
        super().run_command(command)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 986, in run_command
        cmd_obj.run()
      File "/tmp/pip-req-build-qev55xfx/setup.py", line 173, in run
        install.run(self)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/command/install.py", line 68, in run
        return orig.install.run(self)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/command/install.py", line 676, in run
        self.run_command(cmd_name)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/dist.py", line 1214, in run_command
        super().run_command(command)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 984, in run_command
        cmd_obj = self.get_command_obj(command)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 858, in get_command_obj
        klass = self.get_command_class(command)
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/dist.py", line 951, in get_command_class
        self.cmdclass[command] = cmdclass = ep.load()
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_vendor/importlib_metadata/__init__.py", line 194, in load
        module = import_module(match.group('module'))
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/importlib/__init__.py", line 127, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
      File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
      File "<frozen importlib._bootstrap>", line 991, in _find_and_load
      File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 843, in exec_module
      File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
      File "/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/command/install_lib.py", line 4, in <module>
        import distutils.command.install_lib as orig
    ImportError: cannot import name 'install_lib' from 'setuptools._distutils.command' (/home/chrplr/anaconda3/envs/expyriment/lib/python3.8/site-packages/setuptools/_distutils/command/__init__.py)
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-qev55xfx/setup.py", lin
---

**Comment by fladd** [comment]
Hi Christophe,

thanks for the bug report. I just pushed a potential fix to the master branch. Could you try again please and let me know if this worked?

Some background information on the issue as far as I understand it:
`distutils` is deprecated and marked for removal from the standard library. In that context, `setuptools` is supposed to take over its functionality (to support older packages that rely on it), and it uses some hack to shadow the standard lib version (if present) with its own one. For some reason, this does not seem to work. I also realised this recently and attempted to fix this in our setup.py by falling back to direct imports of the internal `distutils` modules that come with `setuptools`. However, this should not be necessary, and I would suggest that you also report this to github.com/pypa/setuptools directly, so they can have a look at it.

One more thing to also check is what version of `distutils` you have currently installed in that venv.

**Comment by chrplr** [comment]
Hi Florian,

I just pulled the latest version of expyriment, but the problem remains (same error message due to `install_lib`).

    distutils.__version__ = 3.9.12
    setuptools..__version__ = 61.2.0

I will report the issue to pypa/setuptools as you suggest.

 (on my Ubuntu 20.04, where setup.py works fine, I have distutils=3.8.8, and  setuptools=52.0.0....)
 
 Oh! Now I realize that I could try to create an venv with an earlier version distutils on my Ubuntu 22 PC... 
 
Thank you!

**Comment by fladd** [comment]
Mmh, that is very strange since I now explicitly try to load something else
in case of the import error you encountered... Are you sure it is the exact
same error you are getting now?

**Comment by chrplr** [comment]
It looks like it is (see below)

---
(base) chrplr@i156478:~/git/expyriment$ git log
commit 4df41a1b1d78f547224ed543653d4122fbca8efa (HEAD -> master, origin/master, origin/HEAD)
Author: Florian Krause <florian.krause@fladd.de>
Date:   Sat May 14 12:17:52 2022 +0200

    Update setup.py

base) chrplr@i156478:~/git/expyriment$ pip install .
Processing /home/chrplr/git/expyriment
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
Requirement already satisfied: pygame<3,>=1.9 in /home/chrplr/anaconda3/lib/python3.9/site-packages (from expyriment===0.10.0-41-g4df41a1) (2.1.2)
Requirement already satisfied: pyopengl<4,>=3.0 in /home/chrplr/anaconda3/lib/python3.9/site-packages (from expyriment===0.10.0-41-g4df41a1) (3.1.6)
Building wheels for collected packages: expyriment
  Building wheel for expyriment (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/chrplr/anaconda3/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-nmsq1vny/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-nmsq1vny/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-0p71z0vh
       cwd: /tmp/pip-req-build-nmsq1vny/
  Complete output (167 lines):
  /home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/dist.py:530: UserWarning: The version specified ('0.10.0-41-g4df41a1') is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
    warnings.warn(
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib
  creating build/lib/expyriment
  copying expyriment/_internals.py -> build/lib/expyriment
  copying expyriment/__init__.py -> build/lib/expyriment
  copying expyriment/cli.py -> build/lib/expyriment
  copying expyriment/_api_reference_tool.py -> build/lib/expyriment
  creating build/lib/expyriment/control
  copying expyriment/control/_miscellaneous.py -> build/lib/expyriment/control
  copying expyriment/control/__init__.py -> build/lib/expyriment/control
  copying expyriment/control/_experiment_control.py -> build/lib/expyriment/control
  copying expyriment/control/_test_suite.py -> build/lib/expyriment/control
  copying expyriment/control/defaults.py -> build/lib/expyriment/control
  creating build/lib/expyriment/io
  copying expyriment/io/_gamepad.py -> build/lib/expyriment/io
  copying expyriment/io/_touchscreenbuttonbox.py -> build/lib/expyriment/io
  copying expyriment/io/_streamingbuttonbox.py -> build/lib/expyriment/io
  copying expyriment/io/_keyboard.py -> build/lib/expyriment/io
  copying expyriment/io/_input_output.py -> build/lib/expyriment/io
  copying expyriment/io/_markeroutput.py -> build/lib/expyriment/io
  copying expyriment/io/_files.py -> build/lib/expyriment/io
  copying expyriment/io/__init__.py -> build/lib/expyriment/io
  copying expyriment/io/_eventbuttonbox.py -> build/lib/expyriment/io
  copying expyriment/io/_textmenu.py -> build/lib/expyriment/io
  copying expyriment/io/_mouse.py -> build/lib/expyriment/io
  copying expyriment/io/_screen.py -> build/lib/expyriment/io
  copying expyriment/io/_textinput.py -> build/lib/expyriment/io
  copying expyriment/io/_serialport.py -> build/lib/expyriment/io
  copying expyriment/io/_triggerinput.py -> build/lib/expyriment/io
  copying expyriment/io/defaults.py -> build/lib/expyriment/io
  creating build/lib/expyriment/io/extras
  copying expyriment/io/extras/__init__.py -> build/lib/expyriment/io/extras
  creating build/lib/expyriment/io/_parallelport
  copying expyriment/io/_parallelport/__init__.py -> build/lib/expyriment/io/_parallelport
  copying expyriment/io/_parallelport/_inpout32.py -> build/lib/expyriment/io/_parallelport
  copying expyriment/io/_parallelport/_dlportio.py -> build/lib/expyriment/io/_parallelport
  copying expyriment/io/_parallelport/_linux.py -> build/lib/expyriment/io/_parallelport
  creating build/lib/expyriment/misc
  copying expyriment/misc/_miscellaneous.py -> build/lib/expyriment/misc
  copying expyriment/misc/_buffer.py -> build/lib/expyriment/misc
  copying expyriment/misc/__init__.py -> build/lib/expyriment/misc
  copying expyriment/misc/_csv_reader_android.py -> build/lib/expyriment/misc
  copying expyriment/misc/_secure_hash.py -> build/lib/expyriment/misc
  copying expyriment/misc/_clock.py -> build/lib/expyriment/misc
  copying expyriment/misc/constants.py -> build/lib/expyriment/misc
  copying expyriment/misc/_timer.py -> build/lib/expyriment/misc
  copying expyriment/misc/_colour.py -> build/lib/expyriment/misc
  copying expyriment/misc/_get_system_info.py -> build/lib/expyriment/misc
  copying expyriment/misc/defaults.py -> build/lib/expyriment/misc
  creating build/lib/expyriment/misc/extras
  copying expyriment/misc/extras/__init__.py -> build/lib/expyriment/misc/extras
  creating build/lib/expyriment/misc/data_preprocessing
  copying expyriment/misc/data_preprocessing/_data_preprocessing.py -> build/lib/expyriment/misc/data_preprocessing
  copying expyriment/misc/data_preprocessing/__init__.py -> build/lib/expyriment/misc/data_preprocessing
  creating build/lib/expyriment/misc/geometry
  copying expyriment/misc/geometry/_geometry.py -> build/lib/expyriment/misc/geometry
  copying expyriment/misc/geometry/_basic_shapes.py -> build/lib/expyriment/misc/geometry
  copying expyriment/misc/geometry/__init__.py -> build/lib/expyriment/misc/geometry
  creating build/lib/expyriment/misc/statistics
  copying expyriment/misc/statistics/_statistics.py -> build/lib/expyriment/misc/statistics
  copying expyriment/misc/statistics/__init__.py -> build/lib/expyriment/misc/statistics
  creating build/lib/expyriment/stimuli
  copying expyriment/stimuli/_tone.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_fixcross.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_textscreen.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_rectangle.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_ellipse.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/__init__.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_visual.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_circle.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_line.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_obsolete.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_audio.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_shape.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_textline.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_stimulus.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_video.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_blankscreen.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_canvas.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_textbox.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/_picture.py -> build/lib/expyriment/stimuli
  copying expyriment/stimuli/defaults.py -> build/lib/expyriment/stimuli
  creating build/lib/expyriment/stimuli/extras
  copying expyriment/stimuli/extras/__init__.py -> build/lib/expyriment/stimuli/extras
  creating build/lib/expyriment/design
  copying expyriment/design/_structure.py -> build/lib/expyriment/design
  copying expyriment/design/__init__.py -> build/lib/expyriment/design
  copying expyriment/design/defaults.py -> build/lib/expyriment/design
  creating build/lib/expyriment/design/extras
  copying expyriment/design/extras/__init__.py -> build/lib/expyriment/design/extras
  creating build/lib/expyriment/design/randomize
  copying expyriment/design/randomize/__init__.py -> build/lib/expyriment/design/randomize
  copying expyriment/design/randomize/_randomize.py -> build/lib/expyriment/design/randomize
  creating build/lib/expyriment/design/permute
  copying expyriment/design/permute/__init__.py -> build/lib/expyriment/design/permute
  copying expyriment/design/permute/_permute.py -> build/lib/expyriment/design/permute
  copying expyriment/expyriment_logo.png -> build/lib/expyriment
  creating build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSerifBold.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeMonoBold.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeMonoOblique.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSansBoldOblique.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSerifItalic.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSerif.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeMonoBoldOblique.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSansOblique.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSans.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSansBold.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeMono.ttf -> build/lib/expyriment/_fonts
  copying expyriment/_fonts/FreeSerifBoldItalic.ttf -> build/lib/expyriment/_fonts
  /home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
    warnings.warn(
  Traceback (most recent call last):
    File "/tmp/pip-req-build-nmsq1vny/setup.py", line 366, in <module>
      run()
    File "/tmp/pip-req-build-nmsq1vny/setup.py", line 315, in run
      setup(name='expyriment',
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/__init__.py", line 87, in setup
      return distutils.core.setup(**attrs)
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 148, in setup
      return run_commands(dist)
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 163, in run_commands
      dist.run_commands()
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 967, in run_commands
      self.run_command(cmd)
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/dist.py", line 1214, in run_command
      super().run_command(command)
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 986, in run_command
      cmd_obj.run()
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/wheel/bdist_wheel.py", line 311, in run
      install_scripts = self.reinitialize_command('install_scripts')
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/__init__.py", line 142, in reinitialize_command
      cmd = _Command.reinitialize_command(self, command, reinit_subcommands)
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 305, in reinitialize_command
      return self.distribution.reinitialize_command(command,
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 939, in reinitialize_command
      command = self.get_command_obj(command_name)
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 858, in get_command_obj
      klass = self.get_command_class(command)
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/dist.py", line 951, in get_command_class
      self.cmdclass[command] = cmdclass = ep.load()
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_vendor/importlib_metadata/__init__.py", line 194, in load
      module = import_module(match.group('module'))
    File "/home/chrplr/anaconda3/lib/python3.9/importlib/__init__.py", line 127, in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
    File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
    File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
    File "<frozen importlib._bootstrap_external>", line 850, in exec_module
    File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
    File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/command/install_scripts.py", line 2, in <module>
      import distutils.command.install_scripts as orig
  ImportError: cannot import name 'install_scripts' from 'setuptools._distutils.command' (/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/command/__init__.py)
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-req-build-nmsq1vny/setup.py", line 369, in <module>
      raise RuntimeError("Building/Installing Expyriment failed!")
  RuntimeError: Building/Installing Expyriment failed!
  ----------------------------------------
  ERROR: Failed building wheel for expyriment
  Running setup.py clean for expyriment
Failed to build expyriment
Installing collected packages: expyriment
    Running setup.py install for expyriment ... error
    ERROR: Command errored out with exit status 1:
     command: /home/chrplr/anaconda3/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-nmsq1vny/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-nmsq1vny/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-j_dl8gsr/install-record.txt --single-version-externally-managed --compile --install-headers /home/chrplr/anaconda3/include/python3.9/expyriment
         cwd: /tmp/pip-req-build-nmsq1vny/
    Complete output (172 lines):
    /home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/dist.py:530: UserWarning: The version specified ('0.10.0-41-g4df41a1') is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
      warnings.warn(
    running install
    /home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
      warnings.warn(
    running build
    running build_py
    creating build
    creating build/lib
    creating build/lib/expyriment
    copying expyriment/_internals.py -> build/lib/expyriment
    copying expyriment/__init__.py -> build/lib/expyriment
    copying expyriment/cli.py -> build/lib/expyriment
    copying expyriment/_api_reference_tool.py -> build/lib/expyriment
    creating build/lib/expyriment/control
    copying expyriment/control/_miscellaneous.py -> build/lib/expyriment/control
    copying expyriment/control/__init__.py -> build/lib/expyriment/control
    copying expyriment/control/_experiment_control.py -> build/lib/expyriment/control
    copying expyriment/control/_test_suite.py -> build/lib/expyriment/control
    copying expyriment/control/defaults.py -> build/lib/expyriment/control
    creating build/lib/expyriment/io
    copying expyriment/io/_gamepad.py -> build/lib/expyriment/io
    copying expyriment/io/_touchscreenbuttonbox.py -> build/lib/expyriment/io
    copying expyriment/io/_streamingbuttonbox.py -> build/lib/expyriment/io
    copying expyriment/io/_keyboard.py -> build/lib/expyriment/io
    copying expyriment/io/_input_output.py -> build/lib/expyriment/io
    copying expyriment/io/_markeroutput.py -> build/lib/expyriment/io
    copying expyriment/io/_files.py -> build/lib/expyriment/io
    copying expyriment/io/__init__.py -> build/lib/expyriment/io
    copying expyriment/io/_eventbuttonbox.py -> build/lib/expyriment/io
    copying expyriment/io/_textmenu.py -> build/lib/expyriment/io
    copying expyriment/io/_mouse.py -> build/lib/expyriment/io
    copying expyriment/io/_screen.py -> build/lib/expyriment/io
    copying expyriment/io/_textinput.py -> build/lib/expyriment/io
    copying expyriment/io/_serialport.py -> build/lib/expyriment/io
    copying expyriment/io/_triggerinput.py -> build/lib/expyriment/io
    copying expyriment/io/defaults.py -> build/lib/expyriment/io
    creating build/lib/expyriment/io/extras
    copying expyriment/io/extras/__init__.py -> build/lib/expyriment/io/extras
    creating build/lib/expyriment/io/_parallelport
    copying expyriment/io/_parallelport/__init__.py -> build/lib/expyriment/io/_parallelport
    copying expyriment/io/_parallelport/_inpout32.py -> build/lib/expyriment/io/_parallelport
    copying expyriment/io/_parallelport/_dlportio.py -> build/lib/expyriment/io/_parallelport
    copying expyriment/io/_parallelport/_linux.py -> build/lib/expyriment/io/_parallelport
    creating build/lib/expyriment/misc
    copying expyriment/misc/_miscellaneous.py -> build/lib/expyriment/misc
    copying expyriment/misc/_buffer.py -> build/lib/expyriment/misc
    copying expyriment/misc/__init__.py -> build/lib/expyriment/misc
    copying expyriment/misc/_csv_reader_android.py -> build/lib/expyriment/misc
    copying expyriment/misc/_secure_hash.py -> build/lib/expyriment/misc
    copying expyriment/misc/_clock.py -> build/lib/expyriment/misc
    copying expyriment/misc/constants.py -> build/lib/expyriment/misc
    copying expyriment/misc/_timer.py -> build/lib/expyriment/misc
    copying expyriment/misc/_colour.py -> build/lib/expyriment/misc
    copying expyriment/misc/_get_system_info.py -> build/lib/expyriment/misc
    copying expyriment/misc/defaults.py -> build/lib/expyriment/misc
    creating build/lib/expyriment/misc/extras
    copying expyriment/misc/extras/__init__.py -> build/lib/expyriment/misc/extras
    creating build/lib/expyriment/misc/data_preprocessing
    copying expyriment/misc/data_preprocessing/_data_preprocessing.py -> build/lib/expyriment/misc/data_preprocessing
    copying expyriment/misc/data_preprocessing/__init__.py -> build/lib/expyriment/misc/data_preprocessing
    creating build/lib/expyriment/misc/geometry
    copying expyriment/misc/geometry/_geometry.py -> build/lib/expyriment/misc/geometry
    copying expyriment/misc/geometry/_basic_shapes.py -> build/lib/expyriment/misc/geometry
    copying expyriment/misc/geometry/__init__.py -> build/lib/expyriment/misc/geometry
    creating build/lib/expyriment/misc/statistics
    copying expyriment/misc/statistics/_statistics.py -> build/lib/expyriment/misc/statistics
    copying expyriment/misc/statistics/__init__.py -> build/lib/expyriment/misc/statistics
    creating build/lib/expyriment/stimuli
    copying expyriment/stimuli/_tone.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_fixcross.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_textscreen.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_rectangle.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_ellipse.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/__init__.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_visual.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_circle.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_line.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_obsolete.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_audio.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_shape.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_textline.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_stimulus.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_video.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_blankscreen.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_canvas.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_textbox.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/_picture.py -> build/lib/expyriment/stimuli
    copying expyriment/stimuli/defaults.py -> build/lib/expyriment/stimuli
    creating build/lib/expyriment/stimuli/extras
    copying expyriment/stimuli/extras/__init__.py -> build/lib/expyriment/stimuli/extras
    creating build/lib/expyriment/design
    copying expyriment/design/_structure.py -> build/lib/expyriment/design
    copying expyriment/design/__init__.py -> build/lib/expyriment/design
    copying expyriment/design/defaults.py -> build/lib/expyriment/design
    creating build/lib/expyriment/design/extras
    copying expyriment/design/extras/__init__.py -> build/lib/expyriment/design/extras
    creating build/lib/expyriment/design/randomize
    copying expyriment/design/randomize/__init__.py -> build/lib/expyriment/design/randomize
    copying expyriment/design/randomize/_randomize.py -> build/lib/expyriment/design/randomize
    creating build/lib/expyriment/design/permute
    copying expyriment/design/permute/__init__.py -> build/lib/expyriment/design/permute
    copying expyriment/design/permute/_permute.py -> build/lib/expyriment/design/permute
    copying expyriment/expyriment_logo.png -> build/lib/expyriment
    creating build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSerifBold.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeMonoBold.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeMonoOblique.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSansBoldOblique.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSerifItalic.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSerif.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeMonoBoldOblique.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSansOblique.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSans.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSansBold.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeMono.ttf -> build/lib/expyriment/_fonts
    copying expyriment/_fonts/FreeSerifBoldItalic.ttf -> build/lib/expyriment/_fonts
    running install_lib
    Traceback (most recent call last):
      File "/tmp/pip-req-build-nmsq1vny/setup.py", line 366, in <module>
        run()
      File "/tmp/pip-req-build-nmsq1vny/setup.py", line 315, in run
        setup(name='expyriment',
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/__init__.py", line 87, in setup
        return distutils.core.setup(**attrs)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 148, in setup
        return run_commands(dist)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 163, in run_commands
        dist.run_commands()
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 967, in run_commands
        self.run_command(cmd)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/dist.py", line 1214, in run_command
        super().run_command(command)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 986, in run_command
        cmd_obj.run()
      File "/tmp/pip-req-build-nmsq1vny/setup.py", line 182, in run
        install.run(self)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/command/install.py", line 68, in run
        return orig.install.run(self)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/command/install.py", line 676, in run
        self.run_command(cmd_name)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/dist.py", line 1214, in run_command
        super().run_command(command)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 984, in run_command
        cmd_obj = self.get_command_obj(command)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 858, in get_command_obj
        klass = self.get_command_class(command)
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/dist.py", line 951, in get_command_class
        self.cmdclass[command] = cmdclass = ep.load()
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_vendor/importlib_metadata/__init__.py", line 194, in load
        module = import_module(match.group('module'))
      File "/home/chrplr/anaconda3/lib/python3.9/importlib/__init__.py", line 127, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
      File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
      File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
      File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 850, in exec_module
      File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
      File "/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/command/install_lib.py", line 4, in <module>
        import distutils.command.install_lib as orig
    ImportError: cannot import name 'install_lib' from 'setuptools._distutils.command' (/home/chrplr/anaconda3/lib/python3.9/site-packages/setuptools/_distutils/command/__init__.py)
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-nmsq1vny/setup.py", line 369, in <module>
        raise RuntimeError("Building/Installing Expyriment failed!")
    RuntimeError: Building/Installing Expyriment failed!
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/chrplr/anaconda3/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-nmsq1vny/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-nmsq1vny/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-j_dl8gsr/install-record.txt --single-version-externally-managed --compile --install-headers /home/chrplr/anaconda3/include/python3.9/expyriment Check the logs for full command output.
(base) chrplr@i156478:~/git/expyriment$ 
---

**Comment by chrplr** [comment]
The crash does not seem to occur at the import time (?). I see no error when I run the three lines:
```python
from setuptools.command.sdist import sdist
from setuptools.command.build_py import build_py
from setuptools.command.install import install
```

 ```console
(base) chrplr@i156478:~/git/expyriment$ git diff HEAD^ HEAD setup.py
diff --git a/setup.py b/setup.py
index 6685351..bdedcb3 100755
--- a/setup.py
+++ b/setup.py
@@ -13,9 +13,18 @@ import sys
 from subprocess import Popen, PIPE, call
 try:
     from setuptools import setup
-    from setuptools.command.sdist import sdist
-    from setuptools.command.build_py import build_py
-    from setuptools.command.install import install
+    try:
+        from setuptools.command.sdist import sdist
+    except ImportError:
+        from setuptools._distutils.command.sdist import sdist
+    try:
+        from setuptools.command.build_py import build_py
+    except ImportError:
+        from setuptools._distutils.command.build_py import build_py
+    try:
+        from setuptools.command.install import install
+    except ImportError:
+        from setuptools._distutils.command.install import install
     try:
         from setuptools.command.install_data import install_data
     except ImportError:
     
(base) chrplr@i156478:~/git/expyriment$ ipython
Python 3.9.12 (main, Apr  5 2022, 06:56:58) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from setuptools.command.sdist import sdist

In [2]: from setuptools.command.build_py import build_py
   ...: 

In [3]: from setuptools.command.install import install
   ...: 

In [4]: from setuptools.command.install import install
   ...: 






```

**Comment by chrplr** [comment]
Hooray! 

Forcing `pip install setuptools=51.0.0` works: it is then possible to build the wheel for expyriment. This solves my problem under Ubuntu 22.04 (but we still have to understand what is going on with setuptools=61.2). 

By the way, I have noticed several times that setuptools do not like the hexa string at the the end of the release string of expyriment: I get the warning:

    WARNING: Built wheel for expyriment is invalid: Metadata 1.2 mandates PEP 440 version, but '0.10.0-41-g4df41a1' is not

Have a good day

**Comment by fladd** [comment]
@chrplr Could you try again? I pushed a fix which at least seems to work on Windows 10, with the newest setuptools.

**Comment by chrplr** [comment]
I just cloned https://github.com/expyriment/expyriment and indeed I can now
install it without the setuptools errors message.
(This under Ubuntu 22.04, with Anaconda Python 3.10.4, setuptools 61.2.0)

```
(test_expyriment) ***@***.***:~/tmp/expyriment$ pip install .
Processing /home/chrplr/tmp/expyriment
  DEPRECATION: A future pip version will change local packages to be built
in-place without first copying to a temporary directory. We recommend you
use --use-feature=in-tree-build to test your packages with this new
behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find
discussion regarding this at https://github.com/pypa/pip/issues/7555.
Collecting pygame<3,>=1.9
  Downloading
pygame-2.1.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
(21.9 MB)
     |████████████████████████████████| 21.9 MB 11.9 MB/s
Collecting pyopengl<4,>=3.0
  Using cached PyOpenGL-3.1.6-py3-none-any.whl (2.4 MB)
Building wheels for collected packages: expyriment
  Building wheel for expyriment (setup.py) ... done
  Created wheel for expyriment:
filename=expyriment-0.10.0+63.g97ebbc5-py3-none-any.whl size=6460704
sha256=3f5e2509fdb485e2841581f7d8d9c772a832d3bb2c235695e8456877a70ede8e
  Stored in directory:
/tmp/pip-ephem-wheel-cache-_idltf7g/wheels/ec/e8/ad/7f59184869782b93ea746b48d186c453ed5ad54333a7c284d0
Successfully built expyriment
Installing collected packages: pyopengl, pygame, expyriment
Successfully installed expyriment-0.10.0+63.g97ebbc5 pygame-2.1.2
pyopengl-3.1.6
```


PS: now, another issue under Ubuntu 22.04, is a run time crash (libGL
error: MESA-LOADER, X error...) when launching `expyriment -T`, probably
because of Wayland. I need to investigate. It is likely not specific to
expyriment.

**Comment by chrplr** [comment]
Thanks a lot Florian!

Indeed, replacing anaconda's libstdc++ by the system's version did the
trick.
After doing this, expyriment runs under Ubuntu 22.04. Kudos !




--
Christophe Pallier (http://www.pallier.org)
INSERM Cognitive Neuroimaging Lab (http://www.unicog.org)

**Comment by fladd** [comment]
@mikila85  The issue is specific to a Python version that is currently not supported by Expyriment. For the current supported Python version (<=3.7) things should just work.

The reason we do not support Python > 3.7 right now has to do with the fact that the package "Pygame" (which we rely on) has been updated to version 2, which, however, includes major low-level changes that heavily affect timing accuracy in Expyriment. The old version of Pygame, is not available from PIP anymore for Python > 3.7.  In addition, the Pygame developers have also split into two different teams over a dispute and as a result there are now two different Pygame packages, each with their own issues. Since timing accuracy is most crucial for Expyriment, we first need these issues to be fixed before we can release a new Expyriment version that is compatible with the new Pygame and hence with newer Python versions. I am currently working on this, in close contact with members of both Pygame teams. Since fixing the remaining Pygame issues is out of my hand, I unfortunately cannot give an estimate on when we are able to release a new Expyriment version. I am sorry for the inconvenience (I am myself also not happy with this situation) and I will keep you updated as soon as there is a solution.

Some context of the current Pygame issues and our interactions with the Pygame teams:
https://github.com/pygame/pygame/pull/3609
https://github.com/pygame/pygame/issues/3619
https://github.com/pygame/pygame/pull/3609

**Comment by chrplr** [comment]
Thanks a lot for the update and the detailed explanations in the github
issues tickets! Is the issue coming from libsdl2 (missing "true fullscreen
mode") or just in pygame? (Is there a way to bypass it with pysdl2, for
example)

**Comment by chrplr** [comment]
I reply to myself after rereading your tickets : this is a pycame2 issue,
sdl2 still allows to use the full screen mode bypassing desktop.

**Comment by fladd** [comment]
This should be solved now with the move to Pygame2

**Comment by berkaykrc** [comment]
this is still an issue while trying to install experiment which is depends on pygame. We tried to use python 3.7 which didn't helped.

**Comment by fladd** [comment]
@berkaykrc The current version of Expyriment (0.10.0) only works with up to Python 3.7 unfortunately, and will create the issues reported in this thread, when attempting to install it on a later Python version. This has to do with Pygame 1.9 not being available anymore from PyPi in later versions, since they moved on to Pygame 2. We are also moving to Pygame 2. This was a slow process due to many issues on the Pygame side, which resulted in improper stimulus presentation timing. Since timing accuracy is of central importance to us, we had to delay the new release for a long time until those issues have been fixed. They now are fixed (mostly at least), and ee are close to a new release of Expyriment which solves the here mentioned issue and allows to be used with a current Python version (e.g. 3.12) again. If you want to test, a pre release is available here: https://github.com/expyriment/expyriment/releases/tag/v0.11.0.dev1
